src/cas/anndata_to_cas.py

```diff
@@ -28,80 +32,105 @@ def anndata2cas(
+        parent_key (str): The key of the parent item.
+        parent_value (Dict[str, Any]): The parent item's dictionary.
+
+    This function modifies `value` to include `parent` (using `parent_key`),
+    `p_accession`, and `parent_rank` based on `parent_value`.
+    """
+    value.update(
+        {
+            "parent": parent_key,
+            "p_accession": parent_value.get("accession"),
+            "parent_rank": parent_value.get("rank"),
+        }
+    )
+
+
+def add_parent_cell_hierarchy(cas: Dict[str, Any], parent_cell_look_up: Dict[str, Any]):
     """
     Adds parent cell hierarchy information to the CAS dictionary.
 
     Args:
-        cas (dict): The CAS dictionary.
-        include_hierarchy (bool): Flag indicating whether to include hierarchy in the output.
-        parent_cell_look_up (dict): Dictionary containing parent cell information.
+        cas (Dict[str, Any]): The CAS dictionary.
+        parent_cell_look_up (Dict[str, Any]): Dictionary containing parent cell information.
 
     Returns:
         None
     """
-    if include_hierarchy:
-        for key, value in parent_cell_look_up.items():
-            for inner_key, inner_value in parent_cell_look_up.items():
-                if value != inner_value and value.get("cell_ids").issubset(
-                    inner_value.get("cell_ids")
-                ):
-                    value.update(
-                        {
-                            "parent": inner_key,
-                            "p_accession": inner_value.get("accession"),
-                        }
-                    )
-
-        annotation_list = cas.get("annotations")
-        for annotation in annotation_list:
-            annotation.update(
-                {
-                    "parent_cell_set_name": parent_cell_look_up.get(
-                        annotation.get("cell_label")
-                    ).get("parent")
-                }
-            )
-            annotation.update(
-                {
-                    "parent_cell_set_accession": parent_cell_look_up.get(
-                        annotation.get("cell_label")
-                    ).get("p_accession")
-                }
-            )
+    for (key, value), (inner_key, inner_value) in itertools.combinations(
+        parent_cell_look_up.items(), 2
+    ):
+        if value.get("parent") is not None and value.get(
+            "parent_rank", 0
+        ) <= inner_value.get("rank"):
+            continue
+
+        if value["cell_ids"] != inner_value["cell_ids"] and value["cell_ids"].issubset(
+            inner_value["cell_ids"]
+        ):
+            update_parent_info(value, inner_key, inner_value)
+        elif value["cell_ids"] == inner_value["cell_ids"]:
+            if int(inner_value["rank"]) < int(value["rank"]):
+                update_parent_info(value, key, value)
+            elif int(value["rank"]) < int(inner_value["rank"]):
+                update_parent_info(value, inner_key, inner_value)
+
```

Comment:
Let's throw an exception ("cell_ids can't be identical at the same rank.") in `else`. This will significantly help us on identifying data related problems.